frame/election-provider-multi-phase/test-election-provider/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+mod mock;
+
+pub(crate) const LOG_TARGET: &str = "tests::epm";
+
+use mock::*;
+use sp_core::Get;
+use sp_npos_elections::{to_supports, StakedAssignment};
+
+use crate::mock::RuntimeOrigin;
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
+			target: crate::LOG_TARGET,
+			concat!("🛠️  ", $patter)  $(, $values)*
+        )
+    };
+}
+
+fn log_current_time() {
+	log!(
+		trace,
+		"block: {:?}, session: {:?}, era: {:?}, EPM phase: {:?} ts: {:?}",
+		System::block_number(),
+		Session::current_index(),
+		Staking::current_era(),
+		ElectionProviderMultiPhase::current_phase(),
+		Timestamp::now()
+	);
+}
+
+#[test]
+/// Replicates the Kusama incident of 8th Dec 2022 and its resolution through the governance
+/// fallback.
+///
+/// After enough slashes to exceed the `Staking::OffendingValidatorsThreshold`, the staking pallet
+/// set `Forcing::ForceNew`. When a new session starts, staking will start to force a new era and
+/// calls <EPM as election_provider>::elect(). If at this point EPM and the staking miners did not
+/// have enough time to queue a new solution (snapshot + solution submission), the election request
+/// fails. If there is no election fallback mechanism in place, EPM enters in emergency mode.
+/// Recovery: Once EPM is in emergency mode, subsequent calls to `elect()` will fail until a new
+/// solution is added to EPM's `QueuedSolution` queue. This can be achieved through
+/// `Call::set_emergency_election_result` or `Call::governance_fallback` dispatchables. Once a new
+/// solution is added to the queue, EPM phase transitions to `Phase::Off` and the election flow
+/// restarts.
+fn enters_emergency_phase_after_forcing_before_elect() {
+	ExtBuilder::default().initialize_first_session(true).build_and_execute(|| {
+		log!(
+			trace,
+			"current validators (staking): {:?}",
+			<Runtime as pallet_staking::SessionInterface<AccountId>>::validators()
+		);
+		let session_validators_before = Session::validators();
+
+		roll_to_epm_off();
+		assert!(ElectionProviderMultiPhase::current_phase().is_off());
+
+		assert_eq!(pallet_staking::ForceEra::<Runtime>::get(), pallet_staking::Forcing::NotForcing);
+		// slashes until staking gets into `Forcing::ForceNew`.
```

Comment:
How is one slash causing? Ideally, we would also test `OffendingValidatorsThreshold` here. So the first slash won't cause `ForceNew`. Second one wont', third one will.

Reply:
Added a `fn slash_through_offending_threshold` helper function to make sure we slash as many validators as needed to cross the offending threshold.